[Bug?]: error sourcemap is wrong

### Duplicates

- [x] I have searched the existing issues

### Latest version

- [x] I have tested the latest version

### Current behavior 😯

<img width="1025" height="772" alt="Image" src="https://github.com/user-attachments/assets/ba57b005-f0b5-4685-a07b-526e00d5a096" />

### Expected behavior 🤔

it should be line 2

### Steps to reproduce 🕹

```tsx
export default function Home() {
  throw new Error("yes");
}
```

### Context 🔦

_No response_

### Your environment 🌎

```shell

```

Let me add some more info to this bug report. I got it by adding a line in `Counter.tsx` in the default app made by `pnpm create solid`, like this:
```ts
export default function Counter() {
  throw new Error("ooo");
  ...
```
As a result, the error page points to a wrong line, but also the error logged to the console, when I click the `Counter.tsx` link in the stacktrace, opens a view of the file with the wrong line marked. There is also one more exception in the console:
```
dev.js:1599 Error: "null" is not in the SourceMap.
    at BasicSourceMapConsumer.SourceMapConsumer_sourceContentFor [as sourceContentFor] (source-map-consumer.js:794:13)
    at Object.fn (createStackFrame.js:44:36)
    at runComputation (dev.js:742:22)
    at updateComputation (dev.js:724:3)
    at runTop (dev.js:833:7)
    at runQueue (dev.js:904:42)
    at completeUpdates (dev.js:860:84)
    at runUpdates (dev.js:850:5)
    at completeLoad (dev.js:330:5)
    at loadEnd (dev.js:325:14)
```

I realised that the file that reportedly originated this error is this: `http://localhost:3000/home/tpreal/git/solid-project/src/components/Counter.tsx`. This includes my local path, and it is wrong. If I throw an error e.g. in a `createEffect`, the file that throw it is `http://localhost:3000/_build/src/components/Counter.tsx`, and then everything, including the source maps, works correctly.

--- a/packages/start/src/shared/dev-overlay/createStackFrame.ts
+++ b/packages/start/src/shared/dev-overlay/createStackFrame.ts
@@ -9,9 +9,20 @@ export interface StackFrameSource {
   column: number;
 }
 
+// Frames from an SSR error carry filesystem paths (or file:// URLs) rather
+// than the http URL the module is served from. Their positions are already
+// mapped back to the original source by the Vite module runner, unlike
+// client frames whose positions refer to the compiled module.
+function isServerSource(path: string): boolean {
+  return !/^https?:\/\//.test(path);
+}
+
 function getActualFileSource(path: string): string {
   if (path.startsWith("file://")) {
-    return "/_build/@fs" + path.substring("file://".length);
+    return "/@fs/" + path.substring("file://".length).replace(/^\/+/, "");
+  }
+  if (isServerSource(path)) {
+    return "/@fs/" + path.replace(/^\/+/, "");
   }
   return path;
 }
@@ -28,16 +39,18 @@ export function createStackFrame(stackframe: StackFrame, isCompiled: () => boole
       if (!source.fileName) {
         return null;
       }
-      const response = await fetch(getActualFileSource(source.fileName));
+      const url = getActualFileSource(source.fileName);
+      const response = await fetch(url);
       if (!response.ok) {
         return null;
       }
       const content = await response.text();
-      const sourceMap = await getSourceMap(source.fileName, content);
+      const sourceMap = await getSourceMap(url, content);
       return {
         source,
         content,
         sourceMap,
+        isServer: isServerSource(source.fileName),
       };
     },
   );
@@ -47,18 +60,36 @@ export function createStackFrame(stackframe: StackFrame, isCompiled: () => boole
     if (!current) {
       return undefined;
     }
-    const { source, content, sourceMap } = current;
+    const { source, content, sourceMap, isServer } = current;
 
     if (!isCompiled() && source.line && source.column && sourceMap) {
-      const result = sourceMap.originalPositionFor({
-        line: source.line,
-        column: source.column,
-      });
-
-      return {
-        ...result,
-        content: sourceMap.sourceContentFor(result.source),
-      } as StackFrameSource;
+      if (isServer) {
+        // The position is already original; only the original content needs
+        // to be pulled out of the source map.
+        const originalContent = sourceMap.sources.length
+          ? sourceMap.sourceContentFor(sourceMap.sources[0]!, true)
+          : null;
+        if (originalContent) {
+          return {
+            source: source.fileName,
+            line: source.line,
+            column: source.column,
+            name: source.functionName,
+            content: originalContent,
+          } as StackFrameSource;
+        }
+      } else {
+        const result = sourceMap.originalPositionFor({
+          line: source.line,
+          column: source.column,
+        });
+        if (result.source) {
+          return {
+            ...result,
+            content: sourceMap.sourceContentFor(result.source, true),
+          } as StackFrameSource;
+        }
+      }
     }
 
     return {

--- a/packages/start/src/shared/dev-overlay/createStackFrame.ts
+++ b/packages/start/src/shared/dev-overlay/createStackFrame.ts
@@ -1,6 +1,9 @@
 import { type Accessor, createMemo, createResource } from "solid-js";
 import getSourceMap from "./get-source-map.ts";
 
+const HTTP_URL_REGEX = /^https?:\/\//;
+const LEADING_SLASH_REGEX = /^\/+/;
+
 export interface StackFrameSource {
   content: string;
   source: string;
@@ -14,15 +17,15 @@ export interface StackFrameSource {
 // mapped back to the original source by the Vite module runner, unlike
 // client frames whose positions refer to the compiled module.
 function isServerSource(path: string): boolean {
-  return !/^https?:\/\//.test(path);
+  return !HTTP_URL_REGEX.test(path);
 }
 
 function getActualFileSource(path: string): string {
   if (path.startsWith("file://")) {
-    return "/@fs/" + path.substring("file://".length).replace(/^\/+/, "");
+    return "/@fs/" + path.substring("file://".length).replace(LEADING_SLASH_REGEX, "");
   }
   if (isServerSource(path)) {
-    return "/@fs/" + path.replace(/^\/+/, "");
+    return "/@fs/" + path.replace(LEADING_SLASH_REGEX, "");
   }
   return path;
 }

